cirq-core/cirq/ops/control_values_test.py

```diff
@@ -14,93 +14,227 @@
-            want = [v1 + v2 for v1 in control_vals1 for v2 in control_vals2]
-            got = [c for c in control_vals1 & control_vals2]
-            eq.add_equality_group(got, want)
-
-    sum_of_products_data = [((1,),), ((0, 1),), ((0, 0), (0, 1), (1, 0))]
+    # 0. Trivial case of 1 control and 1 qubit.
+    eq.add_equality_group(cirq.ProductOfSums([1]), cirq.ProductOfSums(((1,),)))
+    eq.add_equality_group(cirq.ProductOfSums([0]), cirq.ProductOfSums(((0,),)))
+    # 1. Multiple controls for 1 qubit.
+    #   - The number of different "terms" in this case is just 1; since each term corresponds
+    #     to a specific qubit.
+    eq.add_equality_group(cirq.ProductOfSums([[0, 1, 2]]), cirq.ProductOfSums(((0, 1, 2),)))
+    # 2. Multiple qubits, each with 1 control.
+    #   - Duplicates within qubit control values are ignored.
+    #   - Supports initialization by Sequence[int] and Sequence[Collection[int]].
+    #   - Duplicates and Permutation across qubit control value groups leads to different objects.
+    eq.add_equality_group(
+        cirq.ProductOfSums([0, 1, 2]),
+        cirq.ProductOfSums([[0, 0], [1, 1], [2, 2]]),
+        cirq.ProductOfSums([[0], [1], [2]]),
+    )
+    eq.add_equality_group([0, 0, 1, 1, 2, 2])
+    eq.add_equality_group([2, 0, 1])
+    # 3. Multiple controls and multiple qubits.
+    #   - Permutations within qubit control value groups leads to same objects.
+    eq.add_equality_group(
+        cirq.ProductOfSums([(0, 1), (1, 2)]),
+        cirq.ProductOfSums([(1, 0), (2, 1)]),
+        cirq.ProductOfSums([[0, 1], (2, 1)]),
+    )
+    eq.add_equality_group([(1, 2), (0, 1)])
+
+
+def test_init_sum_of_products():
     eq = cirq.testing.EqualsTester()
-    for control_values1 in sum_of_products_data:
-        for control_values2 in sum_of_products_data:
-            control_vals1 = cv.SumOfProducts(control_values1)
-            control_vals2 = cv.SumOfProducts(control_values2)
-            want = [v1 + v2 for v1 in control_vals1 for v2 in control_vals2]
-            got = [c for c in control_vals1 & control_vals2]
-            eq.add_equality_group(got, want)
-
-    pos = cv.ProductOfSums(((1,), (0,)))
-    sop = cv.SumOfProducts(((1, 0), (0, 1)))
-    assert tuple(p for p in pos & sop) == ((1, 0, 1, 0), (1, 0, 0, 1))
-
-    assert tuple(p for p in sop & pos) == ((1, 0, 1, 0), (0, 1, 1, 0))
-
-    with pytest.raises(TypeError):
-        _ = sop & 1
-
-
-def test_and_supported_types():
-    CV = cv.ProductOfSums((1,))
-    with pytest.raises(TypeError):
-        _ = CV & 1
-
+    # 0. Trivial case of 1 control and 1 qubit
+    eq.add_equality_group(cirq.SumOfProducts([1]), cirq.SumOfProducts(((1,),)))
+    eq.add_equality_group(cirq.SumOfProducts([0]), cirq.SumOfProducts(((0,),)))
+    # 1. Multiple controls for 1 qubit.
+    #   - Duplicates and Permutation across different terms is ignored.
+    eq.add_equality_group(
+        cirq.SumOfProducts([0, 1, 2], name="custom name"),
```

Comment:
I think this would cause some ambiguity for `SumOfProducts`. For example:
```python
>>> SumOfProducts([1, 2, 3]) != SumOfProducts([3, 2, 1])
>>> SumOfProducts([[1], [2], [3]]) == SumOfProducts([[3], [2], [1]])
```

The current approach is to assume that the missing axis is always the "inner" axis; so in `SumOfProducts` we add a the `product` axis and in `ProductOfSums` we add the `sum` axis; which avoids inconsistencies of the form described above. 

It also feels natural to me to look at an object and start indexing from the outer-most axis to the inner-most axis; and hence the i'th term of a 1D sum of products representation should represent a "product"; just like it's corresponding 2d version would. 

Do you know of any other such examples in standard libraries where the newly added axis can be inner or outer depending on context / implementation ?

Reply:
Well, in most cases I'd say APIs typically add _outer_ axes rather than inner one, for example this is how numpy broadcasting works (the innermost axes are lined up and then extra outer axes of length 1 are added on the lower-dimensional array to match the higher dimensional one), or a function that expects a sequence lets you specify one item and it will wrap it in a list.

But I still think in this specific scenario it'd make sense to say that we will add the sum axis if needed since that is the one that is "variable" length in that it could apply to the same gate with the same number of qubits and we're just changing the logical predicate. I don't think there's ambiguity with either way of doing things as long as we specify how the constructor works in the docstring (enough to make clear why `SumOfProducts([1, 2, 3]) != SumOfProducts([3, 2, 1])`, which is also the case for `ProductOfSums` btw).

Another option would be to disallow passing in 1D values to the `SumOfProducts` constructor, which I think is how it currently works, and then we could continue the discussion and add that in the future. Then we don't have to hold up the release to decide this.

Reply:
One more thing I'd like to point out is that the API is `Sequence[Union[int, Collection[int]]]` and not `Union[Sequence[int], Sequence[Collection[int]]`. The API is also conveys that "If you specify an int; we'll wrap it in a list". 

In the latter case, we can clearly define that if you provide a 1D object; we'll add an axes for SUM by default; but in the former case I'm not sure what the definition should be.  

For example, How should we interpret `SumOfProducts([1, [2], [3]])` if we always add an axes for SUM? Currently, it will be interpreted as `SumOfProducts([[1], [2], [3]])`. 

Do you suggest that we should change the API to `Union[Sequence[int], Sequence[Collection[int]]` and raise an error if we encounter the case like above?

Reply:
Hmm; this problem doesn't arise with `ProductOfSums`, because the Sum axis is also the inner axis. For consistency, it would make sense to also change the API of `ProductOfSums` and bring the union outside; but that would mean changing the API of all the public methods which used to accept `Sequence[Union[int, Collection[int]]]` historically when we only had product of sums. 

So; `ProductOfSums` would still have `Union[Sequence[int], Collection[Sequence[int]]` but `SumOfProducts` would be `Union[Sequence[int], Collection[Sequence[int]]` ? 

I personally still like the "if you specify an int; we'll wrap it in a list" convention a lot more (and hence keeping things the way they are) but if you have feel very strongly that we should change; I'd be happy to update to the final proposal above (SumOfProducts and ProductOfSums having different API; one supporting inputs of the form `[1, [2], [3]]` and other raising an error).

Reply:
I feel pretty strongly, but I also think this is something that could be added post-1.0 and for now we could keep the implementation of `SumOfProducts` which is documented as accepting `Tuple[Tuple[int, ...], ...]`.

Reply:
Discussed on gchat, we're going to go with `Collection[Sequence[int]]` for now, so has to be specified as 2D. We can discuss making this more flexible in the future.

Reply:
Updated the API. Also; we use a sorted tuple instead of frozenset for `SumOfProducts` internal representation because we want a consistent order of terms in the circuit diagram when plotting controlled operations with sum of products control values.. We could sort the frozen set each time we want to plot the diagram; but that would be wasteful.